[Kok Hai] iP

unfazing opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's, 

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

unfazing committed:
implement vanilla Greet, Echo, Exit

unfazing committed:
implemented vanilla Add, List

unfazing committed:
impemented level 3 without task class

unfazing committed:
implemented Mark as Done with classes

unfazing committed:
implemented ToDos, Events and Deadlines

unfazing committed:
implemented enums, exception handling and added duke command parsers

unfazing committed:
implemented delete task function

unfazing committed:
added automated text UI testing

unfazing committed:
Implement save to text file

unfazing committed:
Implement LocalDate for event and deadline tasks

unfazing committed:
merge branch-Level-8 to master

unfazing committed:
Use more OOP and divide classes into packages

unfazing committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

unfazing committed:
Add gradle support

unfazing committed:
Add a few unit tests for Parser

unfazing committed:
Improve coding style

unfazing committed:
Refactor code and add JavaDoc for Parser and Ui

unfazing committed:
Refactor code and add JavaDoc for task package

unfazing committed:
Add JavaDocs for Storage and DukeException

unfazing committed:
Add JavaDocs for command package

unfazing committed:
Tweak code to comply with coding standard

unfazing committed:
Implement find command

unfazing committed:
Merge branch 'branch-A-CodingStandard'

unfazing committed:
Merge JavaDoc branch to master

unfazing committed:
Fix bug

unfazing committed:
Add JavaDocs

unfazing committed:
Implement Checkstyle and improve coding style

unfazing committed:
Improve coding style

jianminglok commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    public void loadFromSave(TaskList taskList) throws DukeException {
+        File file = new File(SAVE_DIR + SAVE_NAME);
+        ArrayList<Task> taskArray = new ArrayList<>();
```
Maybe you could consider naming this as tasks instead of taskArray?

jianminglok commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    public void saveTaskList(TaskList taskList) {
+        try {
+            boolean dataDir = new File(SAVE_DIR).mkdirs();
```
Maybe you could consider naming this as hasDataDir instead?

jianminglok commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,93 @@
+     * @throws DukeException Thrown when invalid String representation given.
+     */
+    public static Task decode(String str) throws DukeException {
+        String[] splitStr = str.split(" \\| ", 5);
```
Maybe this could be named strSplit instead of splitStr?

jianminglok commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+ */
+
+public class TaskList {
+    protected ArrayList<Task> taskList;
```
Maybe you could consider naming this as tasks instead of taskList?

jianminglok commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,155 @@
+            break;
+        case DEADLINE:
+            // parsed[0] is description of task; parsed[1] is by
+            String[] parsed = Parser.parseDeadlineCommand(userInput);
```
Maybe you could consider naming parsed in a plural form as it is a multi-valued variable?

jianminglok commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,121 @@
+ */
+public class Ui {
+    private int maxCharPerLine = 100;
+    private String horizontal;
```
Maybe you could consider naming this variable differently to make it more clear?

jianminglok commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,121 @@
+     */
+    public void print(String str) {
+        String[] lines = str.split("\n");
+        ArrayList<String> result = new ArrayList<>();
```
Maybe you could consider naming this variable as results?

jianminglok commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private static int parseMarkOrDeleteCommands(String userInput) throws DukeException {
+        String[] splitStr = userInput.split(" ", 2);
```
Maybe this could be named strSplit instead of splitStr?

jianminglok commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private static String parseTodoCommand(String userInput) throws DukeException {
+        String[] splitStr = userInput.split(" ", 2);
```
Maybe this could be named strSplit instead of splitStr?

This is noticed in some other parts of the code as well.

jianminglok commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,155 @@
+            throw new DukeException("duke.Task.Deadline command requires task description and /by argument");
+        } else {
+            String[] output = splitStr[1].split(" /by ", 2);
+            if (output.length < 2) {
```
Maybe this could be named as outputs instead of output?

jianminglok reviewed (commented):
The code is very clean overall and easy to read 👍. Maybe a few minor changes can be made to improve code quality which I have noted down.

jugsliao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+     * Runs the Duke functionality.
+     */
+    public void run() {
+        Scanner s = new Scanner(System.in);
```
Perhaps you could spell out scanner instead of just the letter "s".

jugsliao commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,121 @@
+     * @param taskCount Number of tasks after new task is added.
+     */
+    public void printAddTaskMessage(Task task, int taskCount) {
+        String addTaskMessage = "Got it. I've added this task:\n     " + task //
```
There might be too many white spaces after the "\n".

jugsliao commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,93 @@
+    public static Task decode(String str) throws DukeException {
+        String[] splitStr = str.split(" \\| ", 5);
+        Task result = null;
+        switch(splitStr[0]) {
```
I like how the switch statements has the correct form as it is a common mistake.

jugsliao commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,155 @@
+        String[] splitStr = userInput.split(" ", 2);
+        if (splitStr.length < 2) {
+            throw new DukeException("duke.Task.Event command requires task description, "
+                    + "/from argument and /to argument");
```
I like how you have enough white spaces when splitting your code into two lines.

jugsliao reviewed (commented):
Overall, your code was very clean and easy to read. There were hardly any coding standard violations that I could spot. Good job on writing such good code.

unfazing committed:
Import JavaFX

unfazing committed:
Implement JavaFX

unfazing committed:
Support FXML for JavaFX

unfazing committed:
Update GUI colour and responsive resize

unfazing committed:
Update images and made style changes

unfazing committed:
Fix bug where dialog is truncated

unfazing committed:
Update Duke image

unfazing committed:
Update main class

unfazing committed:
Merge branch 'branch-Level-10'

unfazing committed:
Update GUI colours

unfazing committed:
Refactor all references of "Duke" to "Nemo"

All code refers to application as Duke

In alignment with the new Nemo skin, the application should be renamed

Let's,
* refactor all class names to use "Nemo"
* update all JavaDocs to use "Nemo"

unfazing committed:
Remove comments

unfazing committed:
Removed unnecessary print

unfazing committed:
Add varargs for Find Command

unfazing committed:
Add assertions to TaskList

unfazing committed:
Add assertions to parser

unfazing committed:
Improve code quality

unfazing committed:
Merge pull request #3 from unfazing/branch-A-CodeQuality

Improve code quality

unfazing committed:
Merge branch 'master' into branch-A-Assertions

unfazing committed:
Merge pull request #2 from unfazing/branch-A-Assertions

Add assertions

unfazing committed:
Add Undo and Redo commands

User is unable to undo or redo any changes they make.

It is useful if a user can undo or redo a series of changes if they make
a mistake.

Let's support undo and redo functionality.

unfazing committed:
Highlight errors and changed phrasing

unfazing committed:
Improved checkstyle

unfazing committed:
Include screenshot

unfazing committed:
Fix warnings

unfazing committed:
Update README.md

unfazing committed:
Update README.md

unfazing committed:
Update README.md

unfazing committed:
Fix bugs

unfazing committed:
Merge branch 'master' of https://github.com/unfazing/ip

unfazing committed:
Update README.md

unfazing committed:
Update README.md

unfazing committed:
Create _congif.yml

unfazing committed:
Create config.yml to support emoji

unfazing committed:
Update _config.yml

unfazing committed:
update test

unfazing committed:
remove buggy javafx plugins

unfazing committed:
Update README.md
